oak_idl/src/lib.rs

```diff
@@ -20,120 +20,39 @@ extern crate alloc;
 
 use alloc::vec::Vec;
 use core::fmt::Debug;
-use serde::{Deserialize, Serialize};
 
 pub mod utils;
 
 #[derive(Debug)]
-pub enum ClientError {
+pub enum Error {
+    /// The request message could not be deserialized correctly.
     InvalidRequest,
+    /// The response message could not be serialized correctly.
     InvalidResponse,
-    TransportError(TransportError),
-}
-
-impl From<TransportError> for ClientError {
-    fn from(e: TransportError) -> Self {
-        ClientError::TransportError(e)
-    }
-}
-
-#[derive(Debug)]
-pub enum TransportError {
-    InvalidMessage,
-    InvalidTransactionId,
+    /// The method id provided for the invocation was not implemented by the server.
     InvalidMethodId,
 }
```

Comment:
Should we keep this? I always thought that the seperation between serialization errors (error occurred on the client) and an error response was useful

Reply:
I think we would need to many layers of errors in that case, and there is really not much point for the caller to determine what to do differently in each case (apart from logging something). So, I'm trying to get closer to something like gRPC errors (which are the same on client and server, but some may be generated by the client or server -- see https://grpc.github.io/grpc/core/md_doc_statuscodes.html ). But let's see how it goes, also when we introduce the errors at the framing layer.

Reply:
Alright :). Yh I'm more familiar with HTTP status codes (which are sent by the server only)

Reply:
In that case a different (client specific) error is usually thrown. Eg see the `fetch` API, which returns a JavaScript promise (that can resolve to a value or error). In case that a response is received, the promise succeeds, even if the status code is eg 500 (internal server error). 

In case of a network error, the promise throws a JS error: https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch

Reply:
Which isn't to say that there aren't some client implementations out there that do construct a "fake" status code, but it's not the typical case.